Add type annotations

dmentipl opened the pull request:
This PR adds type annotations to functions and methods, and removes types of arguments to functions or methods in docstrings. 

I have not changed any type annotations in `improver/cli/*.py` as the annotations in those files are "functional", i.e. are used at runtime by clize.

I've added [mypy](https://mypy.readthedocs.io/en/stable/) to the CI (GitHub actions). This ignores the cli files (for the reason discussed above). Given that mypy is currently failing, I have set it so that if mypy fails the whole CI will not fail. Fixing the type issues raised by mypy can be tackled in another PR.

I have added [sphinx_autodoc_typehints](https://pypi.org/project/sphinx-autodoc-typehints/) to automatically read the type annotations and use them for type information in the Sphinx documentation.

This PR changes no *runtime* behaviour. It is only for static testing and building the documentation.

This PR adds two optional (or development) dependencies:

- mypy
- sphinx_autodoc_typehints

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s): added mypy to GitHub actions

dmentipl commented:
For some further background: I used the Dropbox tool [pyannotate](https://github.com/dropbox/pyannotate) (written, in part, by Guido van Rossum) to help generate type annotations. You can run the tool while running the tests to generate type annotations from actually running the code. I found that it added type annotations to about 60% of function/methods. So the rest were generated by hand. Plus I checked over the annotations added automatically. Some required some manual adjustment.

Once this PR is merged I can adjust the code style guide section of the wiki relating to [variable types](https://github.com/metoppv/improver/wiki/Code-Style-Guide#variable-types).

bayliffe commented on improver/nbhood/__init__.py:
```diff
@@ -30,8 +30,12 @@
+
+def radius_by_lead_time(
+    radii: List[str], lead_times: List[str]
+) -> Tuple[Union[float, List[float], Optional[List[int]]]]:
```
It looks as though ```lead_times``` should be optional with a default value of None.

dmentipl replied:
Fixed

bayliffe commented on improver/metadata/amend.py:
```diff
@@ -30,25 +30,26 @@
 
 
-def update_stage_v110_metadata(cube):
+def update_stage_v110_metadata(cube: Cube):
```
```suggestion
def update_stage_v110_metadata(cube: Cube) -> None:
```

dmentipl replied:
Done

bayliffe commented on improver/metadata/amend.py:
```diff
@@ -30,25 +30,26 @@
             Cube to modify attributes in (modified in place)
-
     Returns:
         None
```
We should delete this returns section whilst we are about it.

dmentipl replied:
Done

bayliffe commented on improver/metadata/amend.py:
```diff
@@ -61,14 +62,14 @@ def update_stage_v110_metadata(cube):
 
 
-def amend_attributes(cube, attributes_dict):
+def amend_attributes(cube: Cube, attributes_dict: Dict[str, str]) -> None:
```
```Dict[str, str]```the value portion of the dict is not necessarily a string.

dmentipl replied:
Fixed

bayliffe commented on improver/metadata/amend.py:
```diff
@@ -80,18 +81,18 @@ def amend_attributes(cube, attributes_dict):
 
 
-def set_history_attribute(cube, value, append=False):
+def set_history_attribute(cube: Cube, value: str, append: bool = False):
```
```suggestion
def set_history_attribute(cube: Cube, value: str, append: bool = False) -> None:
```

dmentipl replied:
Done

bayliffe commented on improver/cli/phase_probability.py:
```diff
@@ -41,14 +41,14 @@ def process(*cubes: cli.inputcube, radius: float = 10000.0):
 
     Args:
-        cubes (iris.cube.CubeList or list):
+        cubes:
```
Whilst it is not important, this is the only CLI in which the type declarations have been removed, so maybe put them back so the CLIs remain self-consistent.

dmentipl replied:
This must have been by mistake. I've reverted any changes to `improver/cli/*`.

bayliffe reviewed (commented):
Hey Daniel. This is very far from a complete review as I started it rather late in the day. However, I thought I'd submit what I have found so far in order that you can reply to the questions / suggestions. That will help inform my review of the next 80 files (🥺) as this is not familiar territory for me, so I don't want to keep flagging non-issues. Cheers.

dmentipl commented on improver/between_thresholds.py:
```diff
@@ -45,17 +49,17 @@
     """Calculate the probability of occurrence between thresholds"""
 
-    def __init__(self, threshold_ranges, threshold_units):
+    def __init__(self, threshold_ranges: Any, threshold_units: str) -> None:
```
I suspect this was pyannotate choosing the most general type here. I've fixed it manually.

dmentipl commented on improver/utilities/warnings_handler.py:
```diff
@@ -32,6 +32,9 @@
 import warnings
+from typing import Any, Callable, List
+
+from mypy_extensions import NoReturn
```
Fixed.

dmentipl commented on improver/blending/utilities.py:
```diff
@@ -181,7 +183,7 @@ def _set_blended_time_coords(blended_cube, cycletime):
 
     Args:
         blended_cube (iris.cube.Cube)
```
Done

dmentipl commented on improver/blending/utilities.py:
```diff
@@ -202,20 +204,19 @@ def _set_blended_time_coords(blended_cube, cycletime):
     the cube after blending.
 
     Args:
         blended_cube (iris.cube.Cube)
```
Done

dmentipl commented on improver/utilities/round.py:
```diff
@@ -30,22 +30,26 @@
     data is expected to be very close to the nearest integer.
 
     Args:
         data (np.ndarray of float or float)
```
Done

dmentipl commented on improver/utilities/indexing_operations.py:
```diff
@@ -31,9 +31,10 @@
 
 
-def choose(index_array, array_set):
+def choose(index_array: ndarray, array_set: ndarray) -> ndarray:
```
There is [numpy.typing](https://numpy.org/devdocs/reference/typing.html) but it seems to rely on Python 3.8. Maybe we can come back to this in the future.

dmentipl commented on improver/utilities/cli_utilities.py:
```diff
@@ -31,20 +31,20 @@
 
 
-def load_json_or_none(file_path):
+def load_json_or_none(file_path: str) -> Optional[Dict]:
```
Fixed

dmentipl commented on improver/synthetic_data/utilities.py:
```diff
@@ -78,19 +79,18 @@ def get_leading_dimension(coord_data):
 
 
-def get_height_levels(coord_data):
+def get_height_levels(coord_data: Dict[str, Any]) -> Tuple[List[float], str]:
```
Fixed

dmentipl commented:
@bayliffe Thanks for the review so far! I've carefully checked over the docstrings and function signatures, and removed more types from docstrings. Given the inconsistent nature of hand-written docstrings my regexes didn't pick up everything. I've tried to make the docstrings more consistent. I've also added several missing type annotations.

dmentipl commented:
I think it's probably best to have a separate PR to address the errors that mypy raises.
 (I can raise an issue for this, after this PR is merged.) However, I will note here that running mypy results in the following change:

```
Found 481 errors in 73 files (checked 109 source files)
```

About 80 of those errors relating to `"process" incompatible with supertype "BasePlugin"` can immediately be fixed with the following:

```diff
diff --git a/improver/__init__.py b/improver/__init__.py
index fde9f7fd..68b6a947 100644
--- a/improver/__init__.py
+++ b/improver/__init__.py
@@ -31,6 +31,7 @@
 """Module containing plugin base class."""
 
 from abc import ABC, abstractmethod
+from dask.core import has_tasks
 
 from pkg_resources import DistributionNotFound, get_distribution
 
@@ -57,12 +58,9 @@ class BasePlugin(ABC):
         Returns:
             Output of self.process()
         """
-        return self.process(*args, **kwargs)
-
-    @abstractmethod
-    def process(self, *args, **kwargs):
-        """Abstract class for rest to implement."""
-        pass
+        if hasattr(self, "process"):
+            return self.process(*args, **kwargs)
+        raise NotImplementedError
 
 
 class PostProcessingPlugin(BasePlugin):

```

This also reduces the number of files with errors to 56.

bayliffe commented on improver/utilities/spatial.py:
```diff
@@ -333,47 +332,42 @@ def _create_output_cube(gradient, diff, cube, axis):
+            diff:
                 Cube containing differences along the x or y axis
-            cube (iris.cube.Cube):
+            cube:
                 Cube with correct output dimensions
```
```suggestion
```

bayliffe commented on improver/utilities/spatial.py:
```diff
@@ -476,21 +465,19 @@ def process(self, cube):
 
 
-def lat_lon_determine(cube):
+def lat_lon_determine(cube: Cube) -> Optional[Any]:
```
```suggestion
def lat_lon_determine(cube: Cube) -> Optional[CRS]:
```

bayliffe commented on improver/utilities/temporal.py:
```diff
@@ -341,7 +346,7 @@ def create_output_cube(self, cube, local_time):
         return output_cube
 
-    def _fill_timezones(self, input_cube):
+    def _fill_timezones(self, input_cube: Cube):
```
```suggestion
    def _fill_timezones(self, input_cube: Cube) -> None:
```

bayliffe commented on improver/utilities/temporal.py:
```diff
@@ -386,7 +391,7 @@ def _fill_timezones(self, input_cube):
 
     @staticmethod
-    def _get_time_bounds_offset(input_cube):
+    def _get_time_bounds_offset(input_cube: Cube):
```
```suggestion
    def _get_time_bounds_offset(input_cube: Cube) -> Optional(ndarray):
```

bayliffe commented on improver/utilities/temporal.py:
```diff
@@ -399,7 +404,7 @@ def _get_time_bounds_offset(input_cube):
             return None
 
-    def check_input_cube_dims(self, input_cube, timezone_cube):
+    def check_input_cube_dims(self, input_cube: Cube, timezone_cube: Cube):
```
```suggestion
    def check_input_cube_dims(self, input_cube: Cube, timezone_cube: Cube) -> None:
```

bayliffe commented on improver/utilities/temporal.py:
```diff
@@ -427,7 +432,7 @@ def check_input_cube_dims(self, input_cube, timezone_cube):
             )
 
-    def check_input_cube_time(self, input_cube, local_time):
+    def check_input_cube_time(self, input_cube: Cube, local_time: datetime):
```
```suggestion
    def check_input_cube_time(self, input_cube: Cube, local_time: datetime) -> None:
```

bayliffe commented on improver/utilities/temporal.py:
```diff
@@ -454,7 +459,7 @@ def check_input_cube_time(self, input_cube, local_time):
             )
 
-    def check_timezones_are_unique(self, timezone_cube):
+    def check_timezones_are_unique(self, timezone_cube: Cube):
```
```suggestion
    def check_timezones_are_unique(self, timezone_cube: Cube) -> None:
```

bayliffe commented on improver/utilities/temporal_interpolation.py:
```diff
@@ -55,31 +58,33 @@ class TemporalInterpolation(BasePlugin):
-        self, interval_in_minutes=None, times=None, interpolation_method="linear"
-    ):
+        self,
+        interval_in_minutes: int = None,
+        times: List[datetime] = None,
```
```suggestion
        interval_in_minutes: Optional[int] = None,
        times: Optional[List[datetime]] = None,
```

bayliffe commented on improver/utilities/temporal_interpolation.py:
```diff
@@ -112,24 +117,25 @@ def __repr__(self):
+            initial_time:
                 The start of the period over which a time list is to be
                 constructed.
             final_time (datetime.datetime).
```
Remove type declaration.

bayliffe commented on improver/utilities/warnings_handler.py:
```diff
@@ -41,20 +42,25 @@ class ManageWarnings:
-    def __init__(self, ignored_messages=None, warning_types=None, record=False):
+    def __init__(
+        self,
+        ignored_messages: List[str] = None,
+        warning_types: List[Any] = None,
```
```suggestion
        ignored_messages: Optional[List[str]] = None,
        warning_types: Optional[List[Any]] = None,
```

bayliffe commented on improver/wind_calculations/wind_components.py:
```diff
@@ -139,30 +141,28 @@ def calc_true_north_offset(reference_cube):
-                **v_speed** (iris.cube.Cube):
-                    Cube containing wind vector component in the positive
-                    y-direction
+            - Cube containing wind vector component in the positive
+              x-direction
+            - Cube containing wind vector component in the positive
+              y-direction
```
```suggestion
            - Cube containing wind vector component in the positive
              x-direction u_speed
            - Cube containing wind vector component in the positive
              y-direction v_speed
```
Now I'm just being consistent with my dim recollections of comments from hours ago.... 89/96 files. So close 😄

bayliffe commented on improver/wind_calculations/wind_downscaling.py:
```diff
@@ -397,25 +395,30 @@ def calc_roughness_correction(self, hgrid, uold, mask):
-            u_in (numpy.ndarray):
+            u_in:
                 3D array float32 - velocity on h_in layer, last dim is height
             h_in(numpy.ndarray):
```
Remove type declaration.

bayliffe commented on improver/wind_calculations/wind_downscaling.py:
```diff
@@ -496,27 +499,27 @@ def _interpolate_1d(xup, xlow, at_x, yup, ylow):
     @staticmethod
-    def _interpolate_log(xup, xlow, at_x, yup, ylow):
+    def _interpolate_log(
+        xup: ndarray, xlow: ndarray, at_x: ndarray, yup: ndarray, ylow
```
```suggestion
        xup: ndarray, xlow: ndarray, at_x: ndarray, yup: ndarray, ylow: ndarray
```

bayliffe commented on improver/wind_calculations/wind_downscaling.py:
```diff
@@ -773,29 +768,33 @@ def calc_av_ppgrid_res(self, a_cube):
+    def check_ancils(
+        a_over_s_cube: Cube,
+        sigma_cube: Cube,
+        z0_cube: Cube,
```
```suggestion
        z0_cube: Optional[Cube],
```

bayliffe commented on improver/wxcode/utilities.py:
```diff
@@ -117,16 +116,14 @@ def update_daynight(cubewx):
     """ Update weather cube depending on whether it is day or night
```
Just out of reach of code suggestions. The interface above needs types: `(cubewx: Cube) -> Cube`

dmentipl replied:
I can't add a type annotation here as it causes `test_import_cli` in `improver_tests/cli/test_init.py` to fail.

bayliffe commented on improver/wxcode/utilities.py:
```diff
@@ -218,10 +214,8 @@ def is_variable(thing):
             The string to operate on
```
Remove type declaration.

bayliffe reviewed (changes requested):
@dmentipl this is a massive piece of work, well done for tackling it. Don't be despondent at 75 further comments, that seems like a pretty small number for 4500+ changed lines. I may now need to go and lie down....

I've not yet had a chance to build the sphinx documentation and see how that's working out, so there may be further comments for you on Tuesday (I'm off Friday) if I find issues. We also need to work out the logistics of using mypy at the MO as it is not available by default. That needn't hold up this PR, but is another thing I need to spend time on.

dmentipl commented on improver/__init__.py:
```diff
@@ -76,9 +77,9 @@ def __call__(self, *args, **kwargs):
                 Positional arguments.
             **kwargs:
                 Keyword arguments.
+
```
The issue here is that Cube is yet to be defined so it can't be annotated. The ability to annotate without evaluation was going to be available in Python 3.10 with the adoption of PEP 563 but this has been postponed to Python 3.11.  It is also available with

```python
from __future__ import annotations
```

but this only works for Python 3.7+.

bayliffe replied:
Okay, thanks of explaining.

dmentipl replied:
No worries

dmentipl commented on improver/developer_tools/metadata_interpreter.py:
```diff
@@ -154,8 +157,8 @@ def __init__(self):
-        self.errors = []
-        self.warnings = []
+        self.errors: List[str] = []
+        self.warnings: List[str] = []
```
I put that in when doing some testing of mypy output. I clearly forgot to remove it. It's probably best to remove it for this PR. We can add type annotations to variables as required for fixing the errors raised by mypy in a future PR.

dmentipl commented on improver/utilities/rescale.py:
```diff
@@ -84,8 +92,12 @@ def rescale(data, data_range=None, scale_range=(0.0, 1.0), clip=False):
+    scaled_cube: Cube,
+    data_vals: Tuple[float, float, float],
+    scaling_vals: Tuple[float, float, float],
+    combine_function: Callable[[ndarray, ndarray], ndarray] = np.minimum,
```
Yep, that makes sense.

dmentipl commented:
Thanks @bayliffe very much for the thorough review on this rather large PR!

I've implemented everything suggested with some caveats. I've provided some comments on the individual points of your review. Any point without a comment I implemented exactly as suggested. I've reproduced some of my responses here so they're easier to find.

With regards to `improver/__init__.py`:

> The issue here is that Cube is yet to be defined so it can't be annotated. The ability to annotate without evaluation was going to be available in Python 3.10 with the adoption of PEP 563 but this has been postponed to Python 3.11. It is also available with

```python
from __future__ import annotations
```

> but this only works for Python 3.7+.


With regards to type annotations on variables:

> I put that in when doing some testing of mypy output. I clearly forgot to remove it. It's probably best to remove it for this PR. We can add type annotations to variables as required for fixing the errors raised by mypy in a future PR.

With regards to `improver/wxcode/utilities.py`:

> I can't add a type annotation here as it causes `test_import_cli` in `improver_tests/cli/test_init.py` to fail.

bayliffe reviewed (changes requested):
Thanks for making the second code review so straightforward Daniel. I found one typo, which is no doubt a more pleasing number of comments to receive.

I've noticed the following file in which the doc-strings appear not to have been updated, so those should be done:
- https://github.com/dmentipl/improver/blob/add_type_annotations/improver/nowcasting/forecasting.py

There is another issue that needs to be addressed in this PR, which is the readthedocs documentation environment. This is a lighter-weight environment that is used only to build the [web-based documentation](https://improver.readthedocs.io/en/latest/index.html). This can be found here: https://github.com/metoppv/improver/blob/master/doc/rtd_environment.yml
We need to add the `- sphinx-autodoc-typehints` dependency to this environment. I have tested this on a readthedocs fork and this change is sufficient to get those docs building correctly.

~~I've raised a ticket to fix a related issue that showed up when testing: #1449~~ - My mistake.

dmentipl commented on improver/metadata/forecast_times.py:
```diff
@@ -185,12 +185,13 @@ def _create_frt_type_coord(
+
     Returns:
-        iris.coords.AuxCoord
+        New auxliarry coordinate
```
Done

dmentipl commented:
Many thanks again for the review @bayliffe. I've addressed your comments:

- I added type annotations to `nowcasting/forecast.py`. I initially left them off because the module is deprecated.
- I fixed the typo.
- I added sphinx-autodoc-typehints to the readthedocs environment. Thanks for testing that.

I've also merged in changes from master.

bayliffe commented:
Thanks for capturing the updates from master, if any other changes get merged before this that do not use type-hinting we can add in a fix-up PR. I am happy with this change now. Thanks again for all the work.

bayliffe commented:
I forgot to mention that the code style guide will need to be updated with this change. Please could you make suitable adjustments to this page: https://github.com/metoppv/improver/wiki/Code-Style-Guide

dmentipl commented:
Great, thanks very much @bayliffe.

I've updated the wiki. Perhaps it's time to move the wiki to `doc/` and thus readthedocs?

I'll hold off merging this until @lucyleeow, who has agreed to review, has reviewed.

lucyleeow reviewed (commented):
Not done but just putting this up before I finish up.

 I've mostly just matched type and checked basic logic. I assumed when there is a difference, you've updated to the correct type but I did just commented on some to double check.

dmentipl commented on improver/blending/weighted_blend.py:
```diff
@@ -577,21 +592,21 @@ def get_weights_array(self, cube, weights):
                 in process).
-            weights (iris.cube.Cube):
+            weights:
                 Cube of blending weights.
```
The fact that it's optional is expressed in the type annotation.

dmentipl commented on improver/blending/weighted_blend.py:
```diff
@@ -654,26 +668,25 @@ def weighted_mean(self, cube, weights):
         return result
 
-    def process(self, cube, weights=None):
+    def process(self, cube: Cube, weights: Cube = None) -> Cube:
```
Fixed

dmentipl commented on improver/blending/weights.py:
```diff
@@ -706,59 +703,59 @@ def __repr__(self):
+            width:
+                The width of the triangular function from the centre point.
+            units:
+                The cf units of the width and midpoint.
```
I think it's outside of this PR to adjust the docstrings other than to remove types and make the format appropriate for sphinx-autodoc-typehints.

dmentipl commented on improver/blending/weights.py:
```diff
@@ -231,36 +239,29 @@ def _check_config_dict(self):
+              fill values will be used as constants that are extrapolated
+              if the target_points are outside the source_points
+              provided. These are equal to the first and last values
+              provided by the source weights.
```
I think it's outside of the scope of the PR to make changes to the content of the docstrings. I've put the returns into a list format so that sphinx-autodoc-typehints picks up the types properly.

lucyleeow replied:
Was referring to the headings that were in the old docs (so not changing content), maybe they could be included in the list at the beginning? Also fine if not.

dmentipl commented on improver/calibration/ensemble_calibration.py:
```diff
@@ -635,14 +652,14 @@ class EstimateCoefficientsForEnsembleCalibration(BasePlugin):
+        distribution: str,
+        point_by_point: bool = False,
+        use_default_initial_guess: bool = False,
+        desired_units: Union[str, Unit] = None,
```
I've made it `Optional[Union[str, Unit]]`.

dmentipl commented on improver/calibration/ensemble_calibration.py:
```diff
@@ -922,8 +940,12 @@ def create_coefficients_cubelist(self, optimised_coeffs, historic_forecasts):
+        truths: ndarray,
+        forecast_predictor: ndarray,
+        predictor: str,
+        number_of_realizations: Optional[int],
```
I've made it `int`.

dmentipl commented on improver/calibration/ensemble_calibration.py:
```diff
@@ -1019,13 +1039,13 @@ def compute_initial_guess(
+            cube:
                 A cube to be masked, on the same grid as the landsea_mask.
                 The last two dimensions on this cube must match the dimensions
                 in the landsea_mask cube.
```
Fixed

dmentipl commented on improver/calibration/reliability_calibration.py:
```diff
@@ -531,20 +546,25 @@ def _check_frt_coord(cubes):
 
-    def process(self, cubes, coordinates=None):
+    def process(
+        self, cubes: List[Cube], coordinates: Optional[List[str]] = None
```
Fixed

dmentipl commented on improver/calibration/ensemble_calibration.py:
```diff
@@ -176,30 +187,28 @@ def _calculate_percentage_change_in_last_iteration(self, allvecs):
+        forecast_predictor_data: ndarray,
+        truth_data: ndarray,
+        forecast_var_data: ndarray,
+        sqrt_pi: float,
```
It's fine for type annotations.

dmentipl commented:
Thanks @lucyleeow for the review so far. As a general comment, I think it's outside of the scope of this PR to fix the docstring content, other than removing types and reformatting for sphinx-autodoc-typehints compatibility. Perhaps we could raise an issue to fix some issues with docstrings separately to this PR.

dmentipl commented on improver/calibration/reliability_calibration.py:
```diff
@@ -431,17 +444,18 @@ def process(self, historic_forecasts, truths):
-            iris.cube.CubeList:
-                A cubelist of reliability table cubes, one for each threshold
-                in the historic forecast cubes.
+            A cubelist of reliability table cubes, one for each threshold
```
This returns a merged cube, which I think is a `Cube` not a `CubeList`.

lucyleeow commented on improver/wind_calculations/wind_downscaling.py:
```diff
@@ -644,32 +647,32 @@ class RoughnessCorrection(PostProcessingPlugin):
+        pporo_cube: Cube,
+        modoro_cube: Cube,
+        modres: float,
+        z0_cube: Cube = None,
+        height_levels_cube: Cube = None,
```
optional

lucyleeow commented on improver/wxcode/utilities.py:
```diff
@@ -231,17 +225,16 @@ def is_variable(thing):
+        diagnostic_fields:
 
     Returns:
         list of lists of str
```
remove type annotation..?

lucyleeow commented on improver/wxcode/weather_symbols.py:
```diff
@@ -382,30 +386,35 @@ def construct_extract_constraint(
-            end (int):
+            end:
                 The weather symbol code to which we are tracing all routes.
             omit_nodes (dict) or None:
```
remove type annotation

lucyleeow commented on improver/wxcode/weather_symbols.py:
```diff
@@ -382,30 +386,35 @@ def construct_extract_constraint(
+        start: str,
+        end: int,
+        omit_nodes: Optional[Dict] = None,
+        route: List[str] = None,
```
optional

lucyleeow reviewed (commented):
LGTM!

dmentipl commented on improver/metadata/utilities.py:
```diff
@@ -44,42 +50,41 @@
+    mandatory_attributes: Union[Dict[str, str], LimitedAttributeDict],
+    optional_attributes: Optional[Any] = None,
+    data: Union[MaskedArray, ndarray] = None,
+    dtype: type = np.float32,
```
It's ok to use `type` here. But instead I'll put `Type` which is imported from `typing`. In Python 3.9+ it's possible to use things like `list[str]` instead of `List[str]`. I think using `type` instead of `Type` is similar.

dmentipl commented on improver/nbhood/nbhood.py:
```diff
@@ -63,21 +67,26 @@ class BaseNeighbourhoodProcessing(PostProcessingPlugin):
-    def __init__(self, neighbourhood_method, radii, lead_times=None):
+    def __init__(
+        self,
+        neighbourhood_method,
```
Thanks, will put `Union[CircularNeighbourhood, SquareNeighbourhood]`.

dmentipl commented on improver/synthetic_data/set_up_test_cubes.py:
```diff
@@ -308,22 +343,22 @@ def _construct_dimension_coords(
+    time_bounds: Optional[Tuple[datetime, datetime]] = None,
+    frt: datetime = datetime(2017, 11, 10, 0, 0),
+    realizations: Optional[Union[List[float], ndarray]] = None,
+    include_scalar_coords: Optional[List[Coord]] = None,
```
I think `Coord` is a supertype of `AuxCoord`. I've used `Coord` where either `AuxCoord` or `DimCoord` can go.

dmentipl commented on improver/utilities/mathematical_operations.py:
```diff
@@ -30,9 +30,14 @@
 import iris
 import numpy as np
 import numpy.ma as ma
+from clize.parser import OptionParameter
```
Yes, that's a mistake. The automated tool I used to generate annotations required some manual checking. I missed this one (and `typing.Any`).

dmentipl replied:
@MoseleyS @benfitzpatrick 

With regards to checking for unused imports we could use [flake8](https://github.com/PyCQA/flake8), for example:

```bash
flake8 improver | grep 'imported but unused'
```

However, given that we're only using pylint to catch errors (with the `-E` flag [here](https://github.com/metoppv/improver/blob/ff658db31364c0bdf3af3940736c260e10544705/.github/workflows/tests.yml#L94)), it might be easier to use flake8 generally as our linter rather than pylint. Running flake8 over `improver` and `improver_tests` takes 5 seconds whereas pylint takes 3 min 26 sec (at least on my workstation).

benfitzpatrick replied:
Yep that would be one way to do it! The major attraction of the way we run pylint is the failure on obvious errors - if we could do the same with flake8 that would be good. Not sure how reliable the unused import checking would be, but we could test it. I was actually wondering if our GitHub Hound integration was working correctly - that should be using flake8 and as I understand it should have commented on those PRs...

tjtg replied:
I'm fairly sure that the hound integration is broken - I haven't seen hound comments on PRs for a while. It's unlikely that people have all started writing amazing code and there are no issues being found by hound. The many comments created by hound did used to fill up a lot of vertical space in PR pages.

I'd also be happy for flake8 to get used in the github actions, possibly as a replacement for pylint. I think flake8 could be a better balance of pickiness compared to the current pylint errors -> CI task fails but pylint warnings -> completely ignored approach. The dramatically faster run time of flake8 compared to pylint is also handy.
